madanalysis/multiweight/histogram.py

```diff
@@ -0,0 +1,579 @@
+    def central_sumw_over_events(self) -> float:
+        """Sum of weights over events"""
+        return self.sumw_over_events[0][self.central_idx] - abs(
+            self.sumw_over_events[1][self.central_idx]
+        )
+
+    @property
+    def central_sumw_over_entries(self) -> float:
+        """Sum of weights over entries"""
+        return self.sumw_over_entries[0][self.central_idx] - abs(
+            self.sumw_over_entries[1][self.central_idx]
+        )
+
+    @property
+    def sumw(self) -> float:
+        """sum of weights"""
+        if self.central_sumw_over_entries < 0:
+            return 0.0
+        return self.central_sumw_over_entries
+
+    @property
+    def weights(self) -> np.ndarray:
+        return np.array(
+            [
+                (pos - abs(neg))[self.central_idx]
+                for pos, neg in zip(self._positive_weights, self._negative_weights)
+            ]
+        )
+
+    @property
+    def all_scaled_weights(self):
+        return [
+            (pos - abs(neg)) * self.scale
+            for pos, neg in zip(self._positive_weights, self._negative_weights)
+        ]
+
+    @property
+    def has_scale_unc(self) -> bool:
+        """Does the histogram has scale unc"""
+        return self.weight_collection.has_scale
+
+    @property
+    def scale_uncertainties(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Retreive scale uncertainties
+
+        Returns:
+            ``Tuple[np.ndarray, np.ndarray]``:
+            lower and upper uncertainties per bin
+        """
+        central_weights = self.weights * self.scale
+        bins = self.all_scaled_weights
+
+        if not self.weight_collection.has_scale:
+            central = [b[self.central_idx] for b in bins]
+            return central, central
+
+        weight_loc = (
+            self.weight_collection.get_scale_vars(
+                self.n_point_scale_variation, self.dynamic_scale_choice
+            )
+            .pdfset(self.nominal_weight.pdfset)
+            .loc
+        )
+
+        upper, lower = [], []
+        for current_bin in bins:
+            current_weights = [current_bin[wloc] for wloc in weight_loc]
+            upper.append(max(current_weights))
+            lower.append(min(current_weights))
+        return (
+            np.clip(central_weights - np.array(lower), 0.0, None),
+            np.clip(np.array(upper) - central_weights, 0.0, None),
+        )
+
+    @property
+    def has_pdf_unc(self) -> bool:
+        """Does the histogram has PDFs"""
+        return len(self.weight_collection.pdfsets) > 1
+
+    @property
+    def pdf_uncertainties(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Retreive PDF uncertainties
+
+        Returns:
+            ``Tuple[np.ndarray, np.ndarray]``:
+            lower and upper uncertainties per bin
+        """
+        if not self.has_pdf_unc:
+            return [self.weights * self.scale] * 2
+
+        method = "replicas"  # "eigenvector" #
```

Comment:
@BFuks both implementations are ready to be tested and it can be switched from here (until we hardcode which pdfset uses which method)